thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/TriggerConditionGrouper.java

```diff
@@ -0,0 +1,283 @@
+public class TriggerConditionGrouper implements Grouper<TriggerConditionGrouperSpec> {
+  protected static final Logger LOG = LoggerFactory.getLogger(TriggerConditionGrouper.class);
+
+  private String expression;
+  private String operator;
+  private Map<String, Object> leftOp;
+  private Map<String, Object> rightOp;
+  private InputDataFetcher dataFetcher;
+
+  static final String PROP_DETECTOR_COMPONENT_NAME = "detectorComponentName";
+  static final String PROP_AND = "and";
+  static final String PROP_OR = "or";
+
+  private static final String PROP_OPERATOR = "operator";
+  private static final String PROP_LEFT_OP = "leftOp";
+  private static final String PROP_RIGHT_OP = "rightOp";
+
+  /**
+   * Group based on 'AND' criteria - Entity has anomaly if both sub-entities A and B have anomalies
+   * at the same time. This means we find anomaly overlapping interval.
+   *
+   * Since the anomalies from the respective entities/metrics are merged
+   * before calling the grouper, we do not have to deal with overlapping
+   * anomalies within an entity/metric
+   *
+   * Core logic:
+   * Sort anomalies by start time and iterate through the list
+   * Push first anomaly to stack
+   * For each anomaly following, pop stack and compare with current anomaly
+   * 1. No overlap:
+   *    If the anomaly doesn't overlap, create and push new entity anomaly
+   * 2. Partial overlap:
+   *    If partial overlap, create new grouped entity anomaly and push to stack followed by current anomaly
+   * 3. Full overlap:
+   *    If full overlap, create new grouped entity anomaly and push to stack followed by current anomaly
+   * After the last anomaly, the stack will have one extra anomaly which needs to be popped out.
+   * Now the stack will hold all the grouped/entity anomalies
+   *
+   * Note the each time an entity copy is created, the children needs to be marked and set appropriately
+   *
+   */
+  private List<MergedAnomalyResultDTO> andGrouping(
+      List<MergedAnomalyResultDTO> anomalyListA, List<MergedAnomalyResultDTO> anomalyListB) {
+    Set<MergedAnomalyResultDTO> groupedAnomalies = new HashSet<>();
+    List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
+
+    Stack<MergedAnomalyResultDTO> anomalyStack = new Stack<>();
+    for (MergedAnomalyResultDTO anomaly : anomalies) {
+      if (anomalyStack.isEmpty()) {
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+        continue;
+      }
+
+      MergedAnomalyResultDTO anomalyStackTop = anomalyStack.pop();
+
+      if (anomaly.getStartTime() > anomalyStackTop.getEndTime()) {
+        // No overlap
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+      } else if (anomaly.getEndTime() > anomalyStackTop.getEndTime()) {
+        //Partial overlap
+        anomalyStackTop.setStartTime(anomaly.getStartTime());
+        setEntityChildMapping(anomalyStackTop, anomaly);
```

Comment:
If partial overlap should update the endTime to the minimum of the anomalies.

Reply:
outdated! In or grouping, we find the max anomaly interval coverage.